File: Single Number III/Single_Number_III.cpp
class Solution 
{
public:
    vector<int> singleNumber(vector<int>& nums) 
    {
        // 边界条件处理
        if (nums.size() < 2)
            return vector<int>(0);

        // 第一次遍历
        int res = 0;
        for (auto number : nums)
            res ^= number;
        
        for (int temp = 1; ; )
        {
            if (temp & res)
            {
                res = temp;
                break;
            }
            else
                temp <<= 1;
        }
        
        // 第二次遍历
        int res1 = 0, res2 = 0;
        for (auto number : nums)
        {
            if (res & number)
                res1 ^= number;
            else
                res2 ^= number;
        }

        vector<int> num;
        num.push_back(res1);
        num.push_back(res2);
        return num;
    }
};

File: Single Number III/Single_Number_III.py
# Runtime: 56 ms, faster than 24.38% of Python3 online submissions for Single Number III.
# Memory Usage: 14.4 MB, less than 46.00% of Python3 online submissions for Single Number III.

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        if len(nums) < 2:
            return []
        
        res = 0
        for num in nums:
            res ^= num
        
        temp = 1
        while True:
            if temp & res:
                res = temp
                break
            else:
                temp <<= 1
        
        res1, res2 = 0, 0
        for num in nums:
            if res & num:
                res1 ^= num
            else:
                res2 ^= num
        
        num = []
        num.append(res1)
        num.append(res2)

        return num